src/Block/Types.hs

```diff
@@ -14,24 +14,51 @@ module Block.Types (
   OgmiosResponse (..),
   TxOut (..),
   BlockInfo (..),
+  blockId,
+  blockSlot,
 ) where
 
-import Data.Aeson (FromJSON, ToJSON, withObject, (.:), (.:?))
+import Data.Aeson (FromJSON, ToJSON, Value (Object, String), withObject, withText, (.:), (.:?))
 import Data.Aeson qualified as Aeson
+import Data.ByteString.Lazy (toStrict)
 import Data.HashMap.Strict qualified as HashMap
 import Data.Int (Int64)
 import Data.Map (Map)
 import Data.Map qualified as Map
 import Data.Text (Text)
+import Data.Text.Encoding (decodeUtf8)
 import GHC.Exts (toList)
 import GHC.Generics (Generic)
 
-data BlockInfo = BlockInfo
-  { blockSlot :: Int64
-  , blockId :: Text
-  }
-  deriving stock (Generic, Show)
-  deriving anyclass (ToJSON)
+data BlockInfo = BlockInfo Int64 Text | BlockOrigin
+  deriving stock (Generic, Show, Eq)
+
+instance ToJSON BlockInfo where
+  toJSON BlockOrigin =
+    Aeson.String "origin"
+  toJSON (BlockInfo slot' id') =
+    Aeson.object
+      [ "blockSlot" Aeson..= slot'
+      , "blockId" Aeson..= id'
+      ]
+
+instance FromJSON BlockInfo where
+  -- super ugly
+  parseJSON val =
+    ( case val of
+        (Object _) -> withObject "BlockInfo" $ \object -> BlockInfo <$> (object .: "blockSlot") <*> (object .: "blockId")
+        (String "origin") -> withText "BlockInfo" $ \_ -> return BlockOrigin
+        _ -> error ""
+    )
+      val
```

Comment:
Maybe we use the default implementation for Generic instance here for ToJSON and FromJSON?

That also allows us to use something like this: https://hackage.haskell.org/package/aeson-typescript

Reply:
Yeah I was trying to use Generic instances originally but had to go and write my own since once you stop using record types the format doesn't match what Ogmios wants.